src/Control/Monad/Freer/Interpretation.hs

```diff
@@ -0,0 +1,213 @@
+    :: Member eff' r
+    => (eff ~> eff')
+    -> Eff (eff ': r) ~> Eff r
+natural = naturally id
+{-# INLINE natural #-}
+
+
+------------------------------------------------------------------------------
+-- | Interpret an effect as a monadic action in 'Eff r'.
+interpret :: (eff ~> Eff r) -> Eff (eff ': r) ~> Eff r
+interpret f (Freer m) = Freer $ \k -> m $ \u ->
+  case decomp u of
+    Left x -> k x
+    Right y -> runFreer (f y) k
+{-# INLINE[3] interpret #-}
+
+
+------------------------------------------------------------------------------
+-- | Like 'interpret', but with access to intermediate state.
+stateful
+    :: (eff ~> S.StateT s (Eff r))
+    -> s
+    -> Eff (eff ': r) a -> Eff r (a, s)
+stateful f s = transform (flip S.runStateT s) f
+{-# INLINE stateful #-}
+
+
+------------------------------------------------------------------------------
+-- | Like 'interpret', but with access to intermediate state.
+withStateful
+    :: s
+    -> (eff ~> S.StateT s (Eff r))
+    -> Eff (eff ': r) a -> Eff r (a, s)
+withStateful s f = stateful f s
+{-# INLINE withStateful #-}
+
+
+------------------------------------------------------------------------------
+-- | Replace the topmost layer of the effect stack with another. This is often
+-- useful for interpreters which would like to introduce some intermediate
+-- effects before immediately handling them.
+replace
+    :: (eff1 ~> eff2)
+    -> Eff (eff1 ': r) ~> Eff (eff2 ': r)
+replace = naturally weaken
+{-# INLINE replace #-}
+
+
+------------------------------------------------------------------------------
+-- | Run an effect via the side-effects of a monad transformer.
+transform
+    :: ( MonadTrans t
+       , MFunctor t
+       , forall m. Monad m => Monad (t m)
+       )
+    => (forall m. Monad m => t m a -> m b)
+       -- ^ The strategy for getting out of the monad transformer.
+    -> (eff ~> t (Eff r))
+    -> Eff (eff ': r) a
+    -> Eff r b
+transform lower f (Freer m) = Freer $ \k -> lower $ m $ \u ->
+  case decomp u of
+    Left  x -> lift $ k x
+    Right y -> hoist (usingFreer k) $ f y
+{-# INLINE[3] transform #-}
+
+
+------------------------------------------------------------------------------
+-- | Run an effect, potentially short circuiting in its evaluation.
+shortCircuit
+    :: (eff ~> E.ExceptT e (Eff r))
+    -> Eff (eff ': r) a
+    -> Eff r (Either e a)
+shortCircuit f = transform E.runExceptT f
+{-# INLINE shortCircuit #-}
+
+
+------------------------------------------------------------------------------
+-- | Intercept an effect without removing it from the effect stack.
+intercept
+    :: Member eff r
+    => (eff ~> Eff r)
+    -> Eff r ~> Eff r
+intercept f (Freer m) = Freer $ \k -> m $ \u ->
+  case prj u of
+    Nothing -> k u
+    Just e  -> usingFreer k $ f e
+{-# INLINE intercept #-}
+
+------------------------------------------------------------------------------
+-- | Like 'interpret', but with access to intermediate state.
+interceptS
```

Comment:
I'd like a better way of composing the interpreter combinators. It'd be cool to go `stateful . intercept` rather than needing to write `interceptS` and a bunch of other silly things.